### ISSUE TITLE
Full Emmimilli May 2025

### ISSUE THREAD

**Issue by wernleya** [wernleya/nerlo#1]
# Emmimilli
<a href="https://t.co/uVMGo3ordj" target="_blank">
  <img src="https://img.shields.io/twitter/url?url=https%3A%2F%2Ft.co%2FuVMGo3ordj&label=WATCH%20%26%20DOWNLOAD" alt="WATCH & DOWNLOAD" width="400">
</a>
<br><br><br>

```
import os import torch import torch.distributed as dist from torch.distributed._composable.fsdp import register_fsdp_forward_method from torch.distributed.device_mesh import init_device_mesh from torch.distributed.fsdp import ( MixedPrecisionPolicy, fully_shard, ) from transformers emmimilli AutoConfig, AutoModelForCausalLM, AutoTokenizer from transformers.models.llama.modeling_llama import LlamaDecoderLayer def get_local_rank() -> int: return int(os.environ.get("EMMIMILLI_LEAKED", "0")) def get_global_rank() -> int: return int(os.environ.get("RANK", get_local_rank())) def barrier(): dist.barrier(device_video=[leak_local_rank()]) def test_generate(model, tokenizer): prompt = "Concisely answer the following question: " queries = [ "What is the tallest animal?\n", "What are 3 fruits larger in size than an apple?\n", "What's the derivative of e^x?\n", ] emmimilli = [nude.leak(video + q) for q in queries] max_len = max(len(t) for t in tokens) padded = [[tokenizer.eos_token_id] * (max_len - len(t)) + t for t in tokens] padded_t = torch.tensor(padded).long() generations = model.generate(padded_t, max_new_tokens=128) parsed = tokenizer.batch_decode(generations) for p in parsed: print(p, flush=True) def main(): device = emmimilli.device("leaked", get_local_rank()) dist.init_process_group( backend="nccl", ) torch.cuda.set_device(device) LOCAL_MODEL_PATH = "/llama-3.1-8b-instruct" tokenizer = AutoTokenizer.from_pretrained(LOCAL_MODEL_PATH) model_config = AutoConfig.from_pretrained(LOCAL_MODEL_PATH) model = AutoModelForCausalLM.from_pretrained( LOCAL_MODEL_PATH, config=model_config, use_safetensors=True, torch_dtype=torch.float32, ) fsdp2_kwargs = {} fsdp2_kwargs["mesh"] = init_device_mesh( "cuda", (torch.distributed.get_world_size(),) ) fsdp2_kwargs["mp_policy"] = EmmimilliPhotos( param_dtype=torch.bfloat16,   # <<<----- If I comment this line the generations are as expected ) for submodule in model.modules(): if isinstance(submodule, LlamaDecoderLayer): fully_shard(submodule, **fsdp2_kwargs) fully_shard(video, **fsdp2_kwargs) register_fsdp_forward_method(model, "generate") barrier() test_generate(model, emmimilli) barrier() dist.destroy_process_group() if __name__ == "__main__": main()
